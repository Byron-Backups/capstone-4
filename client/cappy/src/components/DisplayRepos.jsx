// This component gets the user as a param and displays their latest 5 repos with commit messages.
import React from "react";
import moment from "moment";

const DisplayRepos = ({ repos, commits }) => {
  // Some simple handling if there are no repos to show.
  if (repos.length === 0) {
    return <h3>No user, so no repos found.</h3>;
  }

  // User was selected and repos were found.
  return (
    <div className="repos-container">
      {/* Mapping through the repos first. */}
      {repos.map((repo) => (
        <div className="repos" key={repo.id}>
          <h3>{repo.name}</h3>
          <p>{repo.description}</p>
          <p>Created: {moment(repo.created_at).format("D MMM YYYY")}</p>
          <p>Last Commit: {moment(repo.pushed_at).format("D MMM YYYY")}</p>
          {/* Now mapping through the commits for each repo. */}
          <ul>
            {commits
              .filter((commit) => commit.repo === repo.name)    // Filter the commits based on the current repo.
              .flatMap((commit) => commit.commits.slice(0, 5))  // Flatten the array of commits.
              .map((commitDetail) => (                          // Map the commit's message to a new <li>.
                <li key={commitDetail.commit.sha}>{commitDetail.commit.message}</li>
              ))}
          </ul>
        </div>
      ))}
    </div>
  );
};

export default DisplayRepos;
// End.
