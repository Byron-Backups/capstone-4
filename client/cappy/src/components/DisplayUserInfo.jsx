// This component will get the user's data from App.js and display it.
import React from "react";
import moment from "moment";
import userIconImage from "../images/user-icons.jpg";   // stock image for icon.

const DisplayUserInfo = ({user}) => {
  return (
    // This is where we display the user's data, or the default value if there is no user set yet.
    <div className="profile-container">
      <h2>{user.name || "Name"}</h2>
      <img src={user.avatar_url ? user.avatar_url : userIconImage} alt="User Avatar."/>
      <p>@{user.login || "Username"}</p>
      <p>{user.bio || "This profile has no bio."}</p>
      {/* We are using moment to parse the timestamp data and format it. */}
      <p>Joined Github: {moment(user.created_at).format("D MMM YYYY")}</p>
      <p>Repos: {user.public_repos  || "0"}</p>
      <p>Followers: {user.followers || "0"}</p>
      <p>Following: {user.following || "0"}</p>
      <p>Location: {user.location || "Location Not Available."}</p>
      <p>Website: {user.blog || "Website Not Available."}</p>
      <p>Twitter: {user.twitter_username || "Twitter Not Available."}</p>
      <p>Company: {user.company || "Company Not Available."}</p>
    </div>
  );
};

export default DisplayUserInfo;
// End.
