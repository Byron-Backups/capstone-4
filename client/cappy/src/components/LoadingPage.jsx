// This is the loading page that will be displayed while the API's are loading.
import React from "react";
import { Helmet } from "react-helmet";

// It simply shows that the page is loading with a nice wallpaper.
const LoadingPage = () => {
  return (
    <div className="LoadingPage">
      <Helmet>
        <title>Your Page is Loading...</title>
      </Helmet>
      <h1 style={{ textAlign: "center" }}>Loading...</h1>
      {/* Our wallpapers and loading animation are set in app.css. */}
    </div>
  );
};

export default LoadingPage;
// End.
