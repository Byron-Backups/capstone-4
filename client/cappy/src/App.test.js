// Testing file for Cappy.js.

import React from "react";
import { render, screen } from "@testing-library/react";
// Run 'npm i --save-dev jest enzyme'
//import { shallow } from 'enzyme';
// Components.
import LoadingPage from "./components/LoadingPage.jsx";
import DisplayRepos from "./components/DisplayRepos.jsx";
import DisplayUserInfo from "./components/DisplayUserInfo.jsx";

// Mock props:
const mockCommit = [
  {
    repo: "whisper.cpp-cli",
    commits: [
      {
        sha: "c99055cd840dcb20f9b03c2e8941629d9d890111",
        node_id:
          "C_kwDOL0HAqtoAKGM5OTA1NWNkODQwZGNiMjBmOWIwM2MyZTg5NDE2MjlkOWQ4OTAxMTE",
        commit: {
          author: {
            name: "Charlie Marsh",
            email: "charlie.r.marsh@gmail.com",
            date: "2024-05-01T05:03:54Z",
          },
          committer: {
            name: "GitHub",
            email: "noreply@github.com",
            date: "2024-05-01T05:03:54Z",
          },
          message: "Bump version to v0.0.3",
          tree: {
            sha: "4f0f9d40814c37a3e18c626a042fdbbfa833fef6",
            url: "https://api.github.com/repos/charliermarsh/whisper.cpp-cli/git/trees/4f0f9d40814c37a3e18c626a042fdbbfa833fef6",
          },
        },
      },
    ],
  },
];

const mockRepo = [
  {
    id: 792838314,
    node_id: "R_kgDOL0HAqg",
    name: "whisper.cpp-cli",
    description:
      "Packages whisper.cpp into pre-built, pip-installable wheels, for macOS and Linux.",
    created_at: "2024-04-27T17:36:45Z",
    updated_at: "2024-05-24T02:55:54Z",
    pushed_at: "2024-05-01T05:31:27Z",
  },
];

const mockUser = {
  name: "James Bond",
  login: "doubleOh7",
  avatar_url: "https://example.com/007.jpg",
  bio: "Licnese to kill.",
  created_at: "2022-01-01T00:00:00Z",
  public_repos: 10,
  followers: 100,
  following: 13,
  location: "San Francisco, CA",
  blog: "https://bonding.com",
  twitter_username: "doubleOhBehave",
  company: "ACME Inc.",
};

// Test for the loading page.
describe("LoadingPage component", () => {
  it("returns a <div>", () => {
    const component = LoadingPage;
    const { type } = component();
    expect(type).toBe("div");
  });
});

// Test for the DisplayRepos component.
describe("DisplayRepos", () => {
  test("renders no user message if no repos", () => {
    render(<DisplayRepos repos={[]} commits={[]} />);
    const element = screen.getByText(/No user, so no repos found/i);
    expect(element).toBeInTheDocument();
  });
});

// Test for the DisplayRepos component, alt.
describe("DisplayRepos component", () => {
  it("returns a <div> with props", () => {
    // Check for null or undefined props
    expect(mockRepo).toBeTruthy();
    expect(mockCommit).toBeTruthy();

    // Render the component
    render(<DisplayRepos repos={mockRepo} commits={mockCommit} />);

    // Check if the component exists in the document by checking if the repo name & commit have rendered.
    const component = screen.getByText(/whisper.cpp-cli/i);
    expect(component).toBeInTheDocument();
    const component2 = screen.getByText(/Bump version to v0.0.3/i);
    expect(component2).toBeInTheDocument();
  });
});

// Test for the DisplayUserInfo component.
describe("DisplayUserInfo component", () => {
  it("returns a <div> with props", () => {
    // Check for null or undefined props
    expect(mockUser).toBeTruthy();

    // Render the component
    render(<DisplayUserInfo user={mockUser} />);

    // Check if the component exists in the document by checking if the user's name has rendered.
    const component = screen.getByText(/James Bond/i);
    expect(component).toBeInTheDocument();
  });
});

test('MyComponent snapshot test', () => {
  const { asFragment } = render(<DisplayRepos repos={mockRepo} commits={mockCommit} />);
  expect(asFragment()).toMatchSnapshot();
});

// End.