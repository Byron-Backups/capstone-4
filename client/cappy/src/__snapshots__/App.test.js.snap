// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MyComponent snapshot test 1`] = `
<DocumentFragment>
  <div
    class="repos-container"
  >
    <div
      class="repos"
    >
      <h3>
        whisper.cpp-cli
      </h3>
      <p>
        Packages whisper.cpp into pre-built, pip-installable wheels, for macOS and Linux.
      </p>
      <p>
        Created: 27 Apr 2024
      </p>
      <p>
        Last Commit: 30 Apr 2024
      </p>
      <ul>
        <li>
          Bump version to v0.0.3
        </li>
      </ul>
    </div>
  </div>
</DocumentFragment>
`;
